### ISSUE TITLE
Remote lock.json

### ISSUE THREAD

**Issue by samholmes** [denoland/deno#9542]
For the purpose of being able to use deno as an alternative to `curl <url> | bash`, it would be useful to be able to specify `lock.json` files as a URL. This way you can run remote modules with integrity:

```sh
deno --allow-net --lock=https://domain.com/lock.json https://domain.com/mod.ts
```

The threat model here would be measurable to the integrity of both URLs. The benefit here is that given trust for the first URL, we can be confident that the dependencies for mod.ts have not been changed. Without this, we can only guarantee that the contents of mod.ts hasn't change given we trust the domain serving it's contents (the second URL).

Though not perfect, it should be stated, this threat model is a lot better than `curl <url> | bash`. :)

**Comment by WilcoKruijer** [comment]
As we are already trusting the URL we are loading the script from it might also make sense to add support for some kind of header? `X-Deno-Lock: ./lock.json` ?

**Comment by kitsonk** [comment]
No, that would totally undermine the purpose of the lock. They should always be independently asserted.

**Comment by samholmes** [comment]
Might I also add that it'd be useful to use a relative path for the `--lock` argument:

```sh
deno run --lock=./lock.json https://domain.com/mod.ts
```

In this example, the lock file would resolve to `https://domain.com/lock.json`.

This will make it easier to run remote scripts more easily and securely.

**Comment by lucacasonato** [comment]
Relative paths in `--lock` are resolved relative to CWD.

**Comment by samholmes** [comment]
> Relative paths in `--lock` are resolved relative to CWD.

I just realized the flaw in my thinking, and deleted the post right after making it... 

Though, somehow being able to define relative URL lock files would be useful.

**Comment by samholmes** [comment]
Perhaps, if the file doesn't exist locally, and the script is remote, then it could then try getting the lock file from the URL?

**Comment by lucacasonato** [comment]
> Perhaps, if the file doesn't exist locally, and the script is remote, then it could then try getting the lock file from the URL?

That would make our resolution for lock files "magical". Hard no from me.

**Comment by samholmes** [comment]
How can we express that we want to use the lock file local to the script's location and not the call-site? And, why would we want this? The reason for wanting this is that if we run some script using a file path or url, we wouldn't need to change the value for the `--lock` arg: `deno run --lock=lock.json $filePathOrUrl`; in other words, if the script is remote then the lock is remote, if the script is local then the lock is local. I wonder what would be a less magical way to achieve this? Perhaps:

```sh
deno run --lock=lock.json --base-url=https://domain.com/path script.ts
```

This will run "https://domain.com/path/script.ts" using "https://domain.com/path/lock.json" as the lock file. The proposal here is to add a --base-url argument that would allow for us to define the base url or path for _all_ file path arguments.

**Comment by lino-levan** [comment]
I feel as if a remote lock file completely negates the purpose of a lock file. What if someone modified the remote lock file? Unless I'm missing something (which is very possible), there is literally no security one could get using remote lock files over using nothing.

**Comment by samholmes** [comment]
The trust of the remote lock file is on the transport protocol. HTTPS can mutate the remote lock, but IPFS wouldn't be mutable if using a CID address.

**Comment by bartlomieju** [comment]
Discussed this some time ago, the amount of effort required for such a niche use case is not worth it in our eyes.
